File: address-validation-canada-2-nodejs/REST/validate_canadian_municipality_province_rest.js
import axios from 'axios';
import querystring from 'querystring';
import { AVCA2Response } from './avca2_response.js';

/**
 * @constant
 * @type {string}
 * @description The base URL for the live ServiceObjects AVCA2 (Address Validation - Canada) API service.
 */
const LiveBaseUrl = 'https://sws.serviceobjects.com/AVCA2/api.svc/';

/**
 * @constant
 * @type {string}
 * @description The base URL for the backup ServiceObjects AVCA2 API service.
 */
const BackupBaseUrl = 'https://swsbackup.serviceobjects.com/AVCA2/api.svc/';

/**
 * @constant
 * @type {string}
 * @description The base URL for the trial ServiceObjects AVCA2 API service.
 */
const TrialBaseUrl = 'https://trial.serviceobjects.com/AVCA2/api.svc/';

/**
 * @summary
 * Checks if a response from the API is valid by verifying that it either has no Error object
 * or the Error.TypeCode is not equal to '3'.
 * @param {Object} response - The API response object to validate.
 * @returns {boolean} True if the response is valid, false otherwise.
 */
const isValid = (response) => !response?.Error || response.Error.TypeCode !== '3';

/**
 * @summary
 * Constructs a full URL for the ValidateCanadianMunicipalityProvince API endpoint by combining the base URL
 * with query parameters derived from the input parameters.
 * @param {Object} params - An object containing all the input parameters.
 * @param {string} baseUrl - The base URL for the API service (live, backup, or trial).
 * @returns {string} The constructed URL with query parameters.
 */
const buildUrl = (params, baseUrl) =>
    `${baseUrl}ValidateCanadianMunicipalityProvince?${querystring.stringify(params)}`;

/**
 * @summary
 * Performs an HTTP GET request to the specified URL with a given timeout.
 * @param {string} url - The URL to send the GET request to.
 * @param {number} timeoutSeconds - The timeout duration in seconds for the request.
 * @returns {Promise<AVCA2Response>} A promise that resolves to an AVCA2Response object containing the API response data.
 * @throws {Error} Thrown if the HTTP request fails, with a message detailing the error.
 */
const httpGet = async (url, timeoutSeconds) => {
    try {
        const response = await axios.get(url, { timeout: timeoutSeconds * 1000 });
        return new AVCA2Response(response.data);
    } catch (error) {
        throw new Error(`HTTP request failed: ${error.message}`);
    }
};

/**
 * @summary
 * Provides functionality to call the ServiceObjects AVCA2 API's ValidateCanadianMunicipalityProvince endpoint,
 * retrieving validated Canadian municipality and province information with fallback to a backup endpoint for reliability in live mode.
 */
const ValidateCanadianMunicipalityProvinceClient = {
    /**
     * @summary
     * Asynchronously invokes the ValidateCanadianMunicipalityProvince API endpoint, attempting the primary endpoint
     * first and falling back to the backup if the response is invalid (Error.TypeCode == '3') in live mode.
     * @param {string} Municipality - The municipality of the address (e.g., "Toronto"). Optional if postal code is provided.
     * @param {string} Province - The province of the address (e.g., "ON"). Optional if postal code is provided.
     * @param {string} PostalCode - The postal code of the address (e.g., "M5V 2T6"). Optional if municipality and province are provided.
     * @param {string} LicenseKey - Your license key to use the service.
     * @param {boolean} isLive - Value to determine whether to use the live or trial servers.
     * @param {number} timeoutSeconds - Timeout, in seconds, for the call to the service.
     * @returns {Promise<AVCA2Response>} A promise that resolves to an AVCA2Response object.
     */
    async invokeAsync(Municipality, Province, PostalCode, LicenseKey, isLive = true, timeoutSeconds = 15) {
        const params = {
            Municipality,
            Province,
            PostalCode,
            LicenseKey,
            Format: 'JSON'
        };

        const url = buildUrl(params, isLive ? LiveBaseUrl : TrialBaseUrl);
        let response = await httpGet(url, timeoutSeconds);

        if (isLive && !isValid(response)) {
            const fallbackUrl = buildUrl(params, BackupBaseUrl);
            const fallbackResponse = await httpGet(fallbackUrl, timeoutSeconds);
            return fallbackResponse;
        }
        return response;
    },

    /**
     * @summary
     * Synchronously invokes the ValidateCanadianMunicipalityProvince API endpoint by wrapping the async call
     * and awaiting its result immediately.
     * @param {string} Municipality - The municipality of the address (e.g., "Bowen Island"). Optional if postal code is provided.
     * @param {string} Province - The province of the address (e.g., "BC"). Optional if postal code is provided.
     * @param {string} PostalCode - The postal code of the address (e.g., "V0N 1G2"). Optional if municipality and province are provided.
     * @param {string} LicenseKey - Your license key to use the service.
     * @param {boolean} isLive - Value to determine whether to use the live or trial servers.
     * @param {number} timeoutSeconds - Timeout, in seconds, for the call to the service.
     * @returns {AVCA2Response} An AVCA2Response object with validated municipality and province details or an error.
     */
    invoke(Municipality, Province, PostalCode, LicenseKey, isLive = true, timeoutSeconds = 15) {
        return (async () => await this.invokeAsync(
            Municipality, Province, PostalCode, LicenseKey, isLive, timeoutSeconds
        ))();
    }
};

export { ValidateCanadianMunicipalityProvinceClient, AVCA2Response };

File: address-validation-canada-2-python/REST/validate_canadian_municipality_province_rest.py
from avca2_response import AVCA2Response, CanadianAddressInfo, Error
import requests

# Endpoint URLs for ServiceObjects AVCA2 (Address Validation - Canada) API
primary_url = "https://sws.serviceobjects.com/AVCA2/api.svc/ValidateCanadianMunicipalityProvince?"
backup_url = "https://swsbackup.serviceobjects.com/AVCA2/api.svc/ValidateCanadianMunicipalityProvince?"
trial_url = "https://trial.serviceobjects.com/AVCA2/api.svc/ValidateCanadianMunicipalityProvince?"

def validate_canadian_municipality_province(
    municipality: str,
    province: str,
    postal_code: str,
    license_key: str,
    is_live: bool = True,
    timeout_seconds: int = 15
) -> AVCA2Response:
    """
    Call ServiceObjects AVCA2 API's ValidateCanadianMunicipalityProvince endpoint
    to retrieve validated Canadian municipality and province information for a given address.

    Parameters:
        municipality: The municipality of the address (e.g., "Bowen Island"). Optional if postal code is provided.
        province: The province of the address (e.g., "BC"). Optional if postal code is provided.
        postal_code: The postal code of the address (e.g., "V0N 1G2"). Optional if municipality and province are provided.
        license_key: Your ServiceObjects license key.
        is_live: Use live or trial servers.
        timeout_seconds: Timeout for the HTTP request in seconds.

    Returns:
        AVCA2Response: Parsed JSON response with validated address details or error details.

    Raises:
        RuntimeError: If the API returns an error payload.
        requests.RequestException: On network/HTTP failures (trial mode).
    """
    params = {
        "Municipality": municipality,
        "Province": province,
        "PostalCode": postal_code,
        "LicenseKey": license_key,
        "Format":"JSON"
    }
    # Select the base URL: production vs trial
    url = primary_url if is_live else trial_url

    try:
        # Attempt primary (or trial) endpoint
        response = requests.get(url, params=params, timeout=timeout_seconds)
        response.raise_for_status()
        data = response.json()

        # If API returned an error in JSON payload, trigger fallback
        error = data.get('Error')
        if not (error is None or error.get('TypeCode') != "3"):
            if is_live:
                # Try backup URL
                response = requests.get(backup_url, params=params, timeout=timeout_seconds)
                response.raise_for_status()
                data = response.json()

                # If still error, propagate exception
                if 'Error' in data:
                    raise RuntimeError(f"AVCA2 service error: {data['Error']}")
            else:
                # Trial mode error is terminal
                raise RuntimeError(f"AVCA2 trial error: {data['Error']}")

        # Convert JSON response to AVCA2Response for structured access
        error = Error(**data.get("Error", {})) if data.get("Error") else None

        return AVCA2Response(
            CanadianAddressInfo=CanadianAddressInfo(
                Address=data.get("CanadianAddressInfo", {}).get("Address"),
                Address2=data.get("CanadianAddressInfo", {}).get("Address2"),
                Municipality=data.get("CanadianAddressInfo", {}).get("Municipality"),
                Province=data.get("CanadianAddressInfo", {}).get("Province"),
                PostalCode=data.get("CanadianAddressInfo", {}).get("PostalCode"),
                TimeZone=data.get("CanadianAddressInfo", {}).get("TimeZone"),
                AddressNumberFragment=data.get("CanadianAddressInfo", {}).get("AddressNumberFragment"),
                StreetNameFragment=data.get("CanadianAddressInfo", {}).get("StreetNameFragment"),
                StreetTypeFragment=data.get("CanadianAddressInfo", {}).get("StreetTypeFragment"),
                DirectionalCodeFragment=data.get("CanadianAddressInfo", {}).get("DirectionalCodeFragment"),
                UnitTypeFragment=data.get("CanadianAddressInfo", {}).get("UnitTypeFragment"),
                UnitNumberFragment=data.get("CanadianAddressInfo", {}).get("UnitNumberFragment"),
                IsPOBox=data.get("CanadianAddressInfo", {}).get("IsPOBox"),
                BoxNumberFragment=data.get("CanadianAddressInfo", {}).get("BoxNumberFragment"),
                StationInfo=data.get("CanadianAddressInfo", {}).get("StationInfo"),
                DeliveryMode=data.get("CanadianAddressInfo", {}).get("DeliveryMode"),
                DeliveryInstallation=data.get("CanadianAddressInfo", {}).get("DeliveryInstallation")
            ) if data.get("CanadianAddressInfo") else None,
            Error=error,
        )

    except requests.RequestException as req_exc:
        # Network or HTTP-level error occurred
        if is_live:
            try:
                # Fallback to backup URL
                response = requests.get(backup_url, params=params, timeout=timeout_seconds)
                response.raise_for_status()
                data = response.json()
                if "Error" in data:
                    raise RuntimeError(f"AVCA2 backup error: {data['Error']}") from req_exc

                error = Error(**data.get("Error", {})) if data.get("Error") else None

                return AVCA2Response(
                    CanadianAddressInfo=CanadianAddressInfo(
                        Address=data.get("CanadianAddressInfo", {}).get("Address"),
                        Address2=data.get("CanadianAddressInfo", {}).get("Address2"),
                        Municipality=data.get("CanadianAddressInfo", {}).get("Municipality"),
                        Province=data.get("CanadianAddressInfo", {}).get("Province"),
                        PostalCode=data.get("CanadianAddressInfo", {}).get("PostalCode"),
                        TimeZone=data.get("CanadianAddressInfo", {}).get("TimeZone"),
                        AddressNumberFragment=data.get("CanadianAddressInfo", {}).get("AddressNumberFragment"),
                        StreetNameFragment=data.get("CanadianAddressInfo", {}).get("StreetNameFragment"),
                        StreetTypeFragment=data.get("CanadianAddressInfo", {}).get("StreetTypeFragment"),
                        DirectionalCodeFragment=data.get("CanadianAddressInfo", {}).get("DirectionalCodeFragment"),
                        UnitTypeFragment=data.get("CanadianAddressInfo", {}).get("UnitTypeFragment"),
                        UnitNumberFragment=data.get("CanadianAddressInfo", {}).get("UnitNumberFragment"),
                        IsPOBox=data.get("CanadianAddressInfo", {}).get("IsPOBox"),
                        BoxNumberFragment=data.get("CanadianAddressInfo", {}).get("BoxNumberFragment"),
                        StationInfo=data.get("CanadianAddressInfo", {}).get("StationInfo"),
                        DeliveryMode=data.get("CanadianAddressInfo", {}).get("DeliveryMode"),
                        DeliveryInstallation=data.get("CanadianAddressInfo", {}).get("DeliveryInstallation")
                    ) if data.get("CanadianAddressInfo") else None,
                    Error=error,
                )
            except Exception as backup_exc:
                raise RuntimeError("AVCA2 service unreachable on both endpoints") from backup_exc
        else:
            raise RuntimeError(f"AVCA2 trial error: {str(req_exc)}") from req_exc